stan/math/rev/fun/simplex_column_constrain.hpp

```diff
@@ -0,0 +1,173 @@
+#ifndef STAN_MATH_REV_FUN_SIMPLEX_COLUMN_CONSTRAIN_HPP
+#define STAN_MATH_REV_FUN_SIMPLEX_COLUMN_CONSTRAIN_HPP
+
+#include <stan/math/rev/meta.hpp>
+#include <stan/math/rev/core/reverse_pass_callback.hpp>
+#include <stan/math/rev/core/arena_matrix.hpp>
+#include <stan/math/rev/fun/value_of.hpp>
+#include <stan/math/prim/fun/Eigen.hpp>
+#include <stan/math/prim/fun/inv_logit.hpp>
+#include <stan/math/prim/fun/log1p_exp.hpp>
+#include <cmath>
+#include <tuple>
+#include <vector>
+
+namespace stan {
+namespace math {
+
+namespace internal {
+template <typename Mat, int NewOptions, typename = void>
+struct change_eigen_options_impl {};
+
+template <typename Mat, int NewOptions>
+struct change_eigen_options_impl<Mat, NewOptions,
+                                 require_eigen_matrix_base_t<Mat>> {
+  using type
+      = Eigen::Matrix<typename Mat::Scalar, Mat::RowsAtCompileTime,
+                      Mat::ColsAtCompileTime, NewOptions,
+                      Mat::MaxRowsAtCompileTime, Mat::MaxColsAtCompileTime>;
+};
+
+template <typename Mat, int NewOptions>
+struct change_eigen_options_impl<var_value<Mat>, NewOptions,
+                                 require_eigen_matrix_base_t<Mat>> {
+  using type = var_value<Eigen::Matrix<
+      typename Mat::Scalar, Mat::RowsAtCompileTime, Mat::ColsAtCompileTime,
+      NewOptions, Mat::MaxRowsAtCompileTime, Mat::MaxColsAtCompileTime>>;
+};
+
+template <typename Mat, int NewOptions>
+struct change_eigen_options_impl<Mat, NewOptions, require_eigen_array_t<Mat>> {
+  using type
+      = Eigen::Array<typename Mat::Scalar, Mat::RowsAtCompileTime,
+                     Mat::ColsAtCompileTime, NewOptions,
+                     Mat::MaxRowsAtCompileTime, Mat::MaxColsAtCompileTime>;
+};
+
+template <typename Mat, int NewOptions>
+using change_eigen_options_t =
+    typename change_eigen_options_impl<plain_type_t<std::decay_t<Mat>>,
+                                       NewOptions>::type;
+}  // namespace internal
+
+/**
+ * Return a matrix with columns as simplex vectors.
+ * A simplex is a vector containing values greater than or equal
+ * to 0 that sum to 1.  A matrix (K-1, M) unconstrained values
+ * will produce a matrix of simplex vectors of size (K, M).
+ *
+ * The transform is based on a centered stick-breaking process.
+ *
+ * @tparam T Type of matrix to constrain
+ * @param y Free matrix input of dimensionality (K - 1, M)
+ * @return matrix of column simplexes of dimensionality (K, M)
+ */
+template <typename T, require_rev_matrix_t<T>* = nullptr>
+inline auto simplex_column_constrain(const T& y) {
+  using ret_type = plain_type_t<T>;
+  const Eigen::Index N = y.rows();
+  const Eigen::Index M = y.cols();
+  using eigen_mat_rowmajor
+      = Eigen::Matrix<double, Eigen::Dynamic, Eigen::Dynamic, Eigen::RowMajor>;
+  arena_t<eigen_mat_rowmajor> x_val(N + 1, M);
+  if (unlikely(N == 0 || M == 0)) {
+    return ret_type(x_val);
+  }
+  arena_t<internal::change_eigen_options_t<T, Eigen::RowMajor>> arena_y = y;
```

Comment:
Question: is there a cleaner implementation of this?

(It's honestly a question. It seems like you're striving for memory locality by switching to row-major storage. I don't know if there's a better way to accomplish this, but maybe we should try?)

Reply:
Sadly I can't think of one. The problem is that we need to support both types of matrix types, `var<Matrix>` and `Matrix<var>`. So I need some type traits to dig into the matrix and swap out options so we can get a row major matrix here.

And yes the goal here is row major memory. That lets us use simd instructions and keeps the memory we access contiguous